pkg/scheduler/objects/application.go

```diff
@@ -1093,6 +1097,9 @@ func (sa *Application) tryReservedAllocate(headRoom *resources.Resource, nodeIte
 
 		// Do we need a specific node?
 		if ask.GetRequiredNode() != "" {
+			if ask.HasTriggeredPreemption() || time.Since(ask.GetLastPreemptionAttempt()) < sa.preemptionAttemptInterval {
```

Comment:
This check also defends against the scenario when there's no victim and we want to re-attempt preemption (I picked an arbitrary value of 5 seconds). I don't know if this is a legit case or not, if not, then it can be removed.

Reply:
I'm not sure we want to do this. Once we have triggered preemption, the pods will be scheduled for termination, but may not yet have terminated (either gracefully or not). If graceful termination takes longer than 5 seconds, we're going to attempt to trigger again, which could result in evicting even more pods. In the worst case scenario, we could kill everything else running on the node.

Reply:
The is exactly what this change is intended to prevent, ie. once preemption has been triggered for a given ask, it should not be re-tried again - that's why we mark requests. Currently, it's possible that we try to terminate the same set of allocation a dozen times in a second.

Reply:
Ok, I think I understand now.

Reply:
I think @craigcondit was right: once we triggered preemption the kills will happen and cannot be reversed. We do not re-trigger *ever* in any case not even after a short period of time.
The idea is to find a victim list and if we have a list:
-  mark the ask as preemption triggered 
- kill all victims
- mark all victims as preempted

Next scheduling cycle if the ask is still there and has the preemption triggered flag skip it and do not re-trigger. The node is reserved for the ask so it will get scheduled on that node. It could take up to the grace time for the pod to shutdown, which could be loooong in scheduling times.

Reply:
Yes, ensuring above three steps is good enough. If we anticipate any failures after marking victim as preempted (third step), we could break this into 1. victim_marked_for_preemption 2. victim_preempted (after core receive the terminated/killed confirmation from shim). This breakage would help us to retry only the victim staying in "victim_marked_for_preemption" state for so long.

Reply:
The delete of the pod should never fail. If it does there is a problem on the node. A pod that stays in terminating state  for instance will not get removed if we just retry the removal. Sending a second delete will just be ignored by the API server as it is already processing a request like it.